File: src/types/regime.ts
/**
 * Market Regime Types
 * 
 * Multi-dimensional regime tracking for adaptive signal generation
 */

export type TrendRegime = 'strong_up' | 'up' | 'sideways' | 'down' | 'strong_down';
export type VolatilityRegime = 'compressed' | 'normal' | 'elevated' | 'chaotic';
export type LiquidityRegime = 'thin' | 'healthy' | 'heavy';
export type RiskAppetite = 'risk_on' | 'risk_off' | 'rotation';
export type DerivativesRegime = 'short_crowded' | 'long_crowded' | 'balanced';

export interface RegimeDimensions {
  trend: TrendRegime;
  volatility: VolatilityRegime;
  liquidity: LiquidityRegime;
  risk_appetite: RiskAppetite;
  derivatives: DerivativesRegime;
}

export interface MarketRegime {
  dimensions: RegimeDimensions;
  composite: string;
  score: number; // 0-100, higher = more trade-friendly
  timestamp: string;
  
  // Per-dimension scores
  trend_score: number;
  volatility_score: number;
  liquidity_score: number;
  risk_score: number;
  derivatives_score: number;
  
  // Optional dominance data
  btc_dominance?: number;
  usdt_dominance?: number;
  alt_dominance?: number;
}

export interface SymbolRegime {
  symbol: string;
  trend: TrendRegime;
  volatility: VolatilityRegime;
  score: number; // 0-100
}

export interface RegimeMetadata {
  global_regime?: {
    composite: string;
    score: number;
    trend: TrendRegime;
    volatility: VolatilityRegime;
    liquidity: LiquidityRegime;
  };
  symbol_regime?: {
    trend: TrendRegime;
    volatility: VolatilityRegime;
    score: number;
  };
}

// Plan type and conviction
export type PlanType = 'SMC' | 'ATR_FALLBACK' | 'HYBRID';
export type ConvictionClass = 'A' | 'B' | 'C';

// Enhanced scan result with regime + conviction
export interface EnhancedScanResult {
  symbol: string;
  direction: 'LONG' | 'SHORT';
  setup_type: string;
  confidence_score: number;
  risk_reward: number;
  plan_type: PlanType;
  conviction_class: ConvictionClass;
  missing_critical_timeframes: string[];
  regime?: RegimeMetadata;
  
  // Entry/Stop/Targets
  entry_zone: {
    near_entry: number;
    far_entry: number;
    rationale: string;
  };
  stop_loss: {
    level: number;
    distance_atr: number;
    rationale: string;
  };
  targets: Array<{
    level: number;
    percentage: number;
    rationale: string;
  }>;
  
  rationale: string;
  timestamp: string;
}

File: backend/shared/models/regime.py
"""
Market Regime Models - Multi-dimensional regime analysis

Provides structured regime tracking across multiple dimensions:
- Trend (HTF structure)
- Volatility (ATR-based)
- Liquidity (volume-based)
- Risk appetite (dominance-based)
- Derivatives (funding/OI-based)
"""
from dataclasses import dataclass
from typing import Literal, Optional
from datetime import datetime


@dataclass
class RegimeDimensions:
    """Individual regime dimensions"""
    trend: Literal["strong_up", "up", "sideways", "down", "strong_down"]
    volatility: Literal["compressed", "normal", "elevated", "chaotic"]
    liquidity: Literal["thin", "healthy", "heavy"]
    risk_appetite: Literal["risk_on", "risk_off", "rotation"]
    derivatives: Literal["short_crowded", "long_crowded", "balanced"]


@dataclass
class MarketRegime:
    """Complete market regime structure"""
    dimensions: RegimeDimensions
    composite: str  # e.g. "choppy_risk_off", "bullish_risk_on"
    score: float  # 0-100, higher = more trade-friendly
    timestamp: datetime
    
    # Per-dimension scores for visibility
    trend_score: float
    volatility_score: float
    liquidity_score: float
    risk_score: float
    derivatives_score: float
    
    # Context (optional)
    btc_dominance: Optional[float] = None
    usdt_dominance: Optional[float] = None
    alt_dominance: Optional[float] = None


@dataclass
class SymbolRegime:
    """Per-symbol local regime"""
    symbol: str
    trend: Literal["strong_up", "up", "sideways", "down", "strong_down"]
    volatility: Literal["compressed", "normal", "elevated", "chaotic"]
    score: float  # 0-100


@dataclass
class RegimePolicy:
    """Mode-specific regime handling"""
    mode_name: str
    min_regime_score: float
    allow_in_risk_off: bool
    position_size_adjustment: dict  # trend -> multiplier
    confluence_adjustment: dict  # composite_regime -> boost/penalty
    rr_adjustment: dict  # trend/regime -> R:R multiplier
